concordium-cis2/src/cis2_client.rs

```diff
@@ -0,0 +1,433 @@
+        owner: Address,
+        address: Address,
+    ) -> Result<bool, Cis2ClientError> {
+        let params = &OperatorOfQueryParams {
+            queries: vec![OperatorOfQuery {
+                owner,
+                address,
+            }],
+        };
+
+        let is_operators = match host.invoke_contract_read_only(
+            cis_contract_address,
+            params,
+            OPERATOR_OF_ENTRYPOINT_NAME,
+            Amount::from_ccd(0),
+        )? {
+            // Since the contract should return a response. If it doesn't, it is an error.
+            None => bail!(Cis2ClientError::InvokeContractError),
+            Some(mut res) => OperatorOfQueryResponse::deserial(&mut res)?,
+        };
+
+        // If the contract returns a response, but the response is empty, it is an
+        // error. Since for a single query the response should be non-empty.
+        let is_operator = *is_operators.0.first().ok_or(Cis2ClientError::InvokeContractError)?;
+
+        Ok(is_operator)
+    }
+
+    // calls the `balanceOf` entrypoint of the CIS2 contract to get the balance of
+    // the given owner for the given token. Returns the balance of the owner for
+    // the given token.
+    pub fn balance_of<State, S: HasStateApi, T: IsTokenId, A: IsTokenAmount + Copy>(
+        host: &mut impl HasHost<State, StateApiType = S>,
+        cis_contract_address: &ContractAddress,
+        token_id: T,
+        address: Address,
+    ) -> Result<A, Cis2ClientError> {
+        let params = BalanceOfQueryParams {
+            queries: vec![BalanceOfQuery {
+                token_id,
+                address,
+            }],
+        };
+
+        let balances: BalanceOfQueryResponse<A> = match host.invoke_contract_read_only(
+            cis_contract_address,
+            &params,
+            BALANCE_OF_ENTRYPOINT_NAME,
+            Amount::from_ccd(0),
+        )? {
+            // Since the contract should return a response. If it doesn't, it is an error.
+            None => bail!(Cis2ClientError::InvokeContractError),
+            Some(mut res) => BalanceOfQueryResponse::<A>::deserial(&mut res)?,
+        };
+
+        // If the contract returns a response, but the response is empty, it is an
+        // error. Since for a single query the response should be non-empty.
+        let balance = *balances.0.first().ok_or(Cis2ClientError::InvokeContractError)?;
+
+        Ok(balance)
+    }
+
+    // calls the `transfer` entrypoint of the CIS2 contract to transfer the given
+    // amount of tokens from the given owner to the given receiver.
+    // If the transfer is successful, it returns `Ok(())`, else it returns an `Err`.
+    pub fn transfer<State, S: HasStateApi, T: IsTokenId, A: IsTokenAmount + Copy>(
+        host: &mut impl HasHost<State, StateApiType = S>,
+        cis_contract_address: &ContractAddress,
+        token_id: T,
+        amount: A,
+        from: Address,
+        to: Receiver,
```

Comment:
I think it is better to take the transfer directly here
```suggestion
        transfer: Transfer<T, A>,
```

Reply:
done